// @pulumi/kubernetes/core/v1/endpointsPatch.js
"use strict";
// *** WARNING: this file was generated by pulumigen. ***
// *** Do not edit by hand unless you're certain you know what you are doing! ***
Object.defineProperty(exports, "__esModule", { value: true });
exports.EndpointsPatch = void 0;
const pulumi = require("@pulumi/pulumi");
const utilities = require("../../utilities");
/**
 * Patch resources are used to modify existing Kubernetes resources by using
 * Server-Side Apply updates. The name of the resource must be specified, but all other properties are optional. More than
 * one patch may be applied to the same resource, and a random FieldManager name will be used for each Patch resource.
 * Conflicts will result in an error by default, but can be forced using the "pulumi.com/patchForce" annotation. See the
 * [Server-Side Apply Docs](https://www.pulumi.com/registry/packages/kubernetes/how-to-guides/managing-resources-with-server-side-apply/) for
 * additional information about using Server-Side Apply to manage Kubernetes resources with Pulumi.
 * Endpoints is a collection of endpoints that implement the actual service. Example:
 *
 *      Name: "mysvc",
 *      Subsets: [
 *        {
 *          Addresses: [{"ip": "10.10.1.1"}, {"ip": "10.10.2.2"}],
 *          Ports: [{"name": "a", "port": 8675}, {"name": "b", "port": 309}]
 *        },
 *        {
 *          Addresses: [{"ip": "10.10.3.3"}],
 *          Ports: [{"name": "a", "port": 93}, {"name": "b", "port": 76}]
 *        },
 *     ]
 *
 * Endpoints is a legacy API and does not contain information about all Service features. Use discoveryv1.EndpointSlice for complete information about Service endpoints.
 *
 * Deprecated: This API is deprecated in v1.33+. Use discoveryv1.EndpointSlice.
 */
class EndpointsPatch extends pulumi.CustomResource {
    /**
     * Get an existing EndpointsPatch resource's state with the given name, ID, and optional extra
     * properties used to qualify the lookup.
     *
     * @param name The _unique_ name of the resulting resource.
     * @param id The _unique_ provider ID of the resource to lookup.
     * @param opts Optional settings to control the behavior of the CustomResource.
     */
    static get(name, id, opts) {
        return new EndpointsPatch(name, undefined, Object.assign(Object.assign({}, opts), { id: id }));
    }
    /**
     * Returns true if the given object is an instance of EndpointsPatch.  This is designed to work even
     * when multiple copies of the Pulumi SDK have been loaded into the same process.
     */
    static isInstance(obj) {
        if (obj === undefined || obj === null) {
            return false;
        }
        return obj['__pulumiType'] === EndpointsPatch.__pulumiType;
    }
    /**
     * Create a EndpointsPatch resource with the given unique name, arguments, and options.
     *
     * @param name The _unique_ name of the resource.
     * @param args The arguments to use to populate this resource's properties.
     * @param opts A bag of options that control this resource's behavior.
     */
    constructor(name, args, opts) {
        let resourceInputs = {};
        opts = opts || {};
        if (!opts.id) {
            resourceInputs["apiVersion"] = "v1";
            resourceInputs["kind"] = "Endpoints";
            resourceInputs["metadata"] = args ? args.metadata : undefined;
            resourceInputs["subsets"] = args ? args.subsets : undefined;
        }
        else {
            resourceInputs["apiVersion"] = undefined /*out*/;
            resourceInputs["kind"] = undefined /*out*/;
            resourceInputs["metadata"] = undefined /*out*/;
            resourceInputs["subsets"] = undefined /*out*/;
        }
        opts = pulumi.mergeOptions(utilities.resourceOptsDefaults(), opts);
        super(EndpointsPatch.__pulumiType, name, resourceInputs, opts);
    }
}
exports.EndpointsPatch = EndpointsPatch;
/** @internal */
EndpointsPatch.__pulumiType = 'kubernetes:core/v1:EndpointsPatch';
