// @pulumi/kubernetes/resource/v1alpha3/resourceClaimTemplate.js
"use strict";
// *** WARNING: this file was generated by pulumigen. ***
// *** Do not edit by hand unless you're certain you know what you are doing! ***
Object.defineProperty(exports, "__esModule", { value: true });
exports.ResourceClaimTemplate = void 0;
const pulumi = require("@pulumi/pulumi");
const utilities = require("../../utilities");
/**
 * ResourceClaimTemplate is used to produce ResourceClaim objects.
 *
 * This is an alpha type and requires enabling the DynamicResourceAllocation feature gate.
 */
class ResourceClaimTemplate extends pulumi.CustomResource {
    /**
     * Get an existing ResourceClaimTemplate resource's state with the given name, ID, and optional extra
     * properties used to qualify the lookup.
     *
     * @param name The _unique_ name of the resulting resource.
     * @param id The _unique_ provider ID of the resource to lookup.
     * @param opts Optional settings to control the behavior of the CustomResource.
     */
    static get(name, id, opts) {
        return new ResourceClaimTemplate(name, undefined, Object.assign(Object.assign({}, opts), { id: id }));
    }
    /**
     * Returns true if the given object is an instance of ResourceClaimTemplate.  This is designed to work even
     * when multiple copies of the Pulumi SDK have been loaded into the same process.
     */
    static isInstance(obj) {
        if (obj === undefined || obj === null) {
            return false;
        }
        return obj['__pulumiType'] === ResourceClaimTemplate.__pulumiType;
    }
    /**
     * Create a ResourceClaimTemplate resource with the given unique name, arguments, and options.
     *
     * @param name The _unique_ name of the resource.
     * @param args The arguments to use to populate this resource's properties.
     * @param opts A bag of options that control this resource's behavior.
     */
    constructor(name, args, opts) {
        let resourceInputs = {};
        opts = opts || {};
        if (!opts.id) {
            if ((!args || args.spec === undefined) && !opts.urn) {
                throw new Error("Missing required property 'spec'");
            }
            resourceInputs["apiVersion"] = "resource.k8s.io/v1alpha3";
            resourceInputs["kind"] = "ResourceClaimTemplate";
            resourceInputs["metadata"] = args ? args.metadata : undefined;
            resourceInputs["spec"] = args ? args.spec : undefined;
        }
        else {
            resourceInputs["apiVersion"] = undefined /*out*/;
            resourceInputs["kind"] = undefined /*out*/;
            resourceInputs["metadata"] = undefined /*out*/;
            resourceInputs["spec"] = undefined /*out*/;
        }
        opts = pulumi.mergeOptions(utilities.resourceOptsDefaults(), opts);
        const aliasOpts = { aliases: [{ type: "kubernetes:resource.k8s.io/v1alpha1:ResourceClaimTemplate" }, { type: "kubernetes:resource.k8s.io/v1alpha2:ResourceClaimTemplate" }, { type: "kubernetes:resource.k8s.io/v1beta1:ResourceClaimTemplate" }, { type: "kubernetes:resource.k8s.io/v1beta2:ResourceClaimTemplate" }] };
        opts = pulumi.mergeOptions(opts, aliasOpts);
        super(ResourceClaimTemplate.__pulumiType, name, resourceInputs, opts);
    }
}
exports.ResourceClaimTemplate = ResourceClaimTemplate;
/** @internal */
ResourceClaimTemplate.__pulumiType = 'kubernetes:resource.k8s.io/v1alpha3:ResourceClaimTemplate';
