// @pulumi/kubernetes/policy/v1beta1/index.js
"use strict";
// *** WARNING: this file was generated by pulumigen. ***
// *** Do not edit by hand unless you're certain you know what you are doing! ***
Object.defineProperty(exports, "__esModule", { value: true });
exports.PodSecurityPolicyPatch = exports.PodSecurityPolicyList = exports.PodSecurityPolicy = exports.PodDisruptionBudgetPatch = exports.PodDisruptionBudgetList = exports.PodDisruptionBudget = void 0;
const pulumi = require("@pulumi/pulumi");
const utilities = require("../../utilities");
exports.PodDisruptionBudget = null;
utilities.lazyLoad(exports, ["PodDisruptionBudget"], () => require("./podDisruptionBudget"));
exports.PodDisruptionBudgetList = null;
utilities.lazyLoad(exports, ["PodDisruptionBudgetList"], () => require("./podDisruptionBudgetList"));
exports.PodDisruptionBudgetPatch = null;
utilities.lazyLoad(exports, ["PodDisruptionBudgetPatch"], () => require("./podDisruptionBudgetPatch"));
exports.PodSecurityPolicy = null;
utilities.lazyLoad(exports, ["PodSecurityPolicy"], () => require("./podSecurityPolicy"));
exports.PodSecurityPolicyList = null;
utilities.lazyLoad(exports, ["PodSecurityPolicyList"], () => require("./podSecurityPolicyList"));
exports.PodSecurityPolicyPatch = null;
utilities.lazyLoad(exports, ["PodSecurityPolicyPatch"], () => require("./podSecurityPolicyPatch"));
const _module = {
    version: utilities.getVersion(),
    construct: (name, type, urn) => {
        switch (type) {
            case "kubernetes:policy/v1beta1:PodDisruptionBudget":
                return new exports.PodDisruptionBudget(name, undefined, { urn });
            case "kubernetes:policy/v1beta1:PodDisruptionBudgetList":
                return new exports.PodDisruptionBudgetList(name, undefined, { urn });
            case "kubernetes:policy/v1beta1:PodDisruptionBudgetPatch":
                return new exports.PodDisruptionBudgetPatch(name, undefined, { urn });
            case "kubernetes:policy/v1beta1:PodSecurityPolicy":
                return new exports.PodSecurityPolicy(name, undefined, { urn });
            case "kubernetes:policy/v1beta1:PodSecurityPolicyList":
                return new exports.PodSecurityPolicyList(name, undefined, { urn });
            case "kubernetes:policy/v1beta1:PodSecurityPolicyPatch":
                return new exports.PodSecurityPolicyPatch(name, undefined, { urn });
            default:
                throw new Error(`unknown resource type ${type}`);
        }
    },
};
pulumi.runtime.registerResourceModule("kubernetes", "policy/v1beta1", _module);
