// @pulumi/aws/account/index.js
"use strict";
// *** WARNING: this file was generated by pulumi-language-nodejs. ***
// *** Do not edit by hand unless you're certain you know what you are doing! ***
Object.defineProperty(exports, "__esModule", { value: true });
exports.Region = exports.PrimaryContact = exports.getPrimaryContactOutput = exports.getPrimaryContact = exports.AlternativeContact = void 0;
const pulumi = require("@pulumi/pulumi");
const utilities = require("../utilities");
exports.AlternativeContact = null;
utilities.lazyLoad(exports, ["AlternativeContact"], () => require("./alternativeContact"));
exports.getPrimaryContact = null;
exports.getPrimaryContactOutput = null;
utilities.lazyLoad(exports, ["getPrimaryContact", "getPrimaryContactOutput"], () => require("./getPrimaryContact"));
exports.PrimaryContact = null;
utilities.lazyLoad(exports, ["PrimaryContact"], () => require("./primaryContact"));
exports.Region = null;
utilities.lazyLoad(exports, ["Region"], () => require("./region"));
const _module = {
    version: utilities.getVersion(),
    construct: (name, type, urn) => {
        switch (type) {
            case "aws:account/alternativeContact:AlternativeContact":
                return new exports.AlternativeContact(name, undefined, { urn });
            case "aws:account/primaryContact:PrimaryContact":
                return new exports.PrimaryContact(name, undefined, { urn });
            case "aws:account/region:Region":
                return new exports.Region(name, undefined, { urn });
            default:
                throw new Error(`unknown resource type ${type}`);
        }
    },
};
pulumi.runtime.registerResourceModule("aws", "account/alternativeContact", _module);
pulumi.runtime.registerResourceModule("aws", "account/primaryContact", _module);
pulumi.runtime.registerResourceModule("aws", "account/region", _module);
