// @pulumi/docker/remoteImage.js
"use strict";
// *** WARNING: this file was generated by pulumi-language-nodejs. ***
// *** Do not edit by hand unless you're certain you know what you are doing! ***
Object.defineProperty(exports, "__esModule", { value: true });
exports.RemoteImage = void 0;
const pulumi = require("@pulumi/pulumi");
const utilities = require("./utilities");
class RemoteImage extends pulumi.CustomResource {
    /**
     * Get an existing RemoteImage resource's state with the given name, ID, and optional extra
     * properties used to qualify the lookup.
     *
     * @param name The _unique_ name of the resulting resource.
     * @param id The _unique_ provider ID of the resource to lookup.
     * @param state Any extra arguments used during the lookup.
     * @param opts Optional settings to control the behavior of the CustomResource.
     */
    static get(name, id, state, opts) {
        return new RemoteImage(name, state, Object.assign(Object.assign({}, opts), { id: id }));
    }
    /**
     * Returns true if the given object is an instance of RemoteImage.  This is designed to work even
     * when multiple copies of the Pulumi SDK have been loaded into the same process.
     */
    static isInstance(obj) {
        if (obj === undefined || obj === null) {
            return false;
        }
        return obj['__pulumiType'] === RemoteImage.__pulumiType;
    }
    constructor(name, argsOrState, opts) {
        let resourceInputs = {};
        opts = opts || {};
        if (opts.id) {
            const state = argsOrState;
            resourceInputs["build"] = state ? state.build : undefined;
            resourceInputs["forceRemove"] = state ? state.forceRemove : undefined;
            resourceInputs["imageId"] = state ? state.imageId : undefined;
            resourceInputs["keepLocally"] = state ? state.keepLocally : undefined;
            resourceInputs["name"] = state ? state.name : undefined;
            resourceInputs["platform"] = state ? state.platform : undefined;
            resourceInputs["pullTriggers"] = state ? state.pullTriggers : undefined;
            resourceInputs["repoDigest"] = state ? state.repoDigest : undefined;
            resourceInputs["triggers"] = state ? state.triggers : undefined;
        }
        else {
            const args = argsOrState;
            if ((!args || args.name === undefined) && !opts.urn) {
                throw new Error("Missing required property 'name'");
            }
            resourceInputs["build"] = args ? args.build : undefined;
            resourceInputs["forceRemove"] = args ? args.forceRemove : undefined;
            resourceInputs["keepLocally"] = args ? args.keepLocally : undefined;
            resourceInputs["name"] = args ? args.name : undefined;
            resourceInputs["platform"] = args ? args.platform : undefined;
            resourceInputs["pullTriggers"] = args ? args.pullTriggers : undefined;
            resourceInputs["triggers"] = args ? args.triggers : undefined;
            resourceInputs["imageId"] = undefined /*out*/;
            resourceInputs["repoDigest"] = undefined /*out*/;
        }
        opts = pulumi.mergeOptions(utilities.resourceOptsDefaults(), opts);
        super(RemoteImage.__pulumiType, name, resourceInputs, opts);
    }
}
exports.RemoteImage = RemoteImage;
/** @internal */
RemoteImage.__pulumiType = 'docker:index/remoteImage:RemoteImage';
