// @pulumi/kubernetes/storage/v1alpha1/index.js
"use strict";
// *** WARNING: this file was generated by pulumigen. ***
// *** Do not edit by hand unless you're certain you know what you are doing! ***
Object.defineProperty(exports, "__esModule", { value: true });
exports.VolumeAttributesClassPatch = exports.VolumeAttributesClassList = exports.VolumeAttributesClass = exports.VolumeAttachmentPatch = exports.VolumeAttachmentList = exports.VolumeAttachment = void 0;
const pulumi = require("@pulumi/pulumi");
const utilities = require("../../utilities");
exports.VolumeAttachment = null;
utilities.lazyLoad(exports, ["VolumeAttachment"], () => require("./volumeAttachment"));
exports.VolumeAttachmentList = null;
utilities.lazyLoad(exports, ["VolumeAttachmentList"], () => require("./volumeAttachmentList"));
exports.VolumeAttachmentPatch = null;
utilities.lazyLoad(exports, ["VolumeAttachmentPatch"], () => require("./volumeAttachmentPatch"));
exports.VolumeAttributesClass = null;
utilities.lazyLoad(exports, ["VolumeAttributesClass"], () => require("./volumeAttributesClass"));
exports.VolumeAttributesClassList = null;
utilities.lazyLoad(exports, ["VolumeAttributesClassList"], () => require("./volumeAttributesClassList"));
exports.VolumeAttributesClassPatch = null;
utilities.lazyLoad(exports, ["VolumeAttributesClassPatch"], () => require("./volumeAttributesClassPatch"));
const _module = {
    version: utilities.getVersion(),
    construct: (name, type, urn) => {
        switch (type) {
            case "kubernetes:storage.k8s.io/v1alpha1:VolumeAttachment":
                return new exports.VolumeAttachment(name, undefined, { urn });
            case "kubernetes:storage.k8s.io/v1alpha1:VolumeAttachmentList":
                return new exports.VolumeAttachmentList(name, undefined, { urn });
            case "kubernetes:storage.k8s.io/v1alpha1:VolumeAttachmentPatch":
                return new exports.VolumeAttachmentPatch(name, undefined, { urn });
            case "kubernetes:storage.k8s.io/v1alpha1:VolumeAttributesClass":
                return new exports.VolumeAttributesClass(name, undefined, { urn });
            case "kubernetes:storage.k8s.io/v1alpha1:VolumeAttributesClassList":
                return new exports.VolumeAttributesClassList(name, undefined, { urn });
            case "kubernetes:storage.k8s.io/v1alpha1:VolumeAttributesClassPatch":
                return new exports.VolumeAttributesClassPatch(name, undefined, { urn });
            default:
                throw new Error(`unknown resource type ${type}`);
        }
    },
};
pulumi.runtime.registerResourceModule("kubernetes", "storage.k8s.io/v1alpha1", _module);
