// @pulumi/kubernetes/admissionregistration/v1beta1/index.js
"use strict";
// *** WARNING: this file was generated by pulumigen. ***
// *** Do not edit by hand unless you're certain you know what you are doing! ***
Object.defineProperty(exports, "__esModule", { value: true });
exports.ValidatingWebhookConfigurationPatch = exports.ValidatingWebhookConfigurationList = exports.ValidatingWebhookConfiguration = exports.ValidatingAdmissionPolicyPatch = exports.ValidatingAdmissionPolicyList = exports.ValidatingAdmissionPolicyBindingPatch = exports.ValidatingAdmissionPolicyBindingList = exports.ValidatingAdmissionPolicyBinding = exports.ValidatingAdmissionPolicy = exports.MutatingWebhookConfigurationPatch = exports.MutatingWebhookConfigurationList = exports.MutatingWebhookConfiguration = void 0;
const pulumi = require("@pulumi/pulumi");
const utilities = require("../../utilities");
exports.MutatingWebhookConfiguration = null;
utilities.lazyLoad(exports, ["MutatingWebhookConfiguration"], () => require("./mutatingWebhookConfiguration"));
exports.MutatingWebhookConfigurationList = null;
utilities.lazyLoad(exports, ["MutatingWebhookConfigurationList"], () => require("./mutatingWebhookConfigurationList"));
exports.MutatingWebhookConfigurationPatch = null;
utilities.lazyLoad(exports, ["MutatingWebhookConfigurationPatch"], () => require("./mutatingWebhookConfigurationPatch"));
exports.ValidatingAdmissionPolicy = null;
utilities.lazyLoad(exports, ["ValidatingAdmissionPolicy"], () => require("./validatingAdmissionPolicy"));
exports.ValidatingAdmissionPolicyBinding = null;
utilities.lazyLoad(exports, ["ValidatingAdmissionPolicyBinding"], () => require("./validatingAdmissionPolicyBinding"));
exports.ValidatingAdmissionPolicyBindingList = null;
utilities.lazyLoad(exports, ["ValidatingAdmissionPolicyBindingList"], () => require("./validatingAdmissionPolicyBindingList"));
exports.ValidatingAdmissionPolicyBindingPatch = null;
utilities.lazyLoad(exports, ["ValidatingAdmissionPolicyBindingPatch"], () => require("./validatingAdmissionPolicyBindingPatch"));
exports.ValidatingAdmissionPolicyList = null;
utilities.lazyLoad(exports, ["ValidatingAdmissionPolicyList"], () => require("./validatingAdmissionPolicyList"));
exports.ValidatingAdmissionPolicyPatch = null;
utilities.lazyLoad(exports, ["ValidatingAdmissionPolicyPatch"], () => require("./validatingAdmissionPolicyPatch"));
exports.ValidatingWebhookConfiguration = null;
utilities.lazyLoad(exports, ["ValidatingWebhookConfiguration"], () => require("./validatingWebhookConfiguration"));
exports.ValidatingWebhookConfigurationList = null;
utilities.lazyLoad(exports, ["ValidatingWebhookConfigurationList"], () => require("./validatingWebhookConfigurationList"));
exports.ValidatingWebhookConfigurationPatch = null;
utilities.lazyLoad(exports, ["ValidatingWebhookConfigurationPatch"], () => require("./validatingWebhookConfigurationPatch"));
const _module = {
    version: utilities.getVersion(),
    construct: (name, type, urn) => {
        switch (type) {
            case "kubernetes:admissionregistration.k8s.io/v1beta1:MutatingWebhookConfiguration":
                return new exports.MutatingWebhookConfiguration(name, undefined, { urn });
            case "kubernetes:admissionregistration.k8s.io/v1beta1:MutatingWebhookConfigurationList":
                return new exports.MutatingWebhookConfigurationList(name, undefined, { urn });
            case "kubernetes:admissionregistration.k8s.io/v1beta1:MutatingWebhookConfigurationPatch":
                return new exports.MutatingWebhookConfigurationPatch(name, undefined, { urn });
            case "kubernetes:admissionregistration.k8s.io/v1beta1:ValidatingAdmissionPolicy":
                return new exports.ValidatingAdmissionPolicy(name, undefined, { urn });
            case "kubernetes:admissionregistration.k8s.io/v1beta1:ValidatingAdmissionPolicyBinding":
                return new exports.ValidatingAdmissionPolicyBinding(name, undefined, { urn });
            case "kubernetes:admissionregistration.k8s.io/v1beta1:ValidatingAdmissionPolicyBindingList":
                return new exports.ValidatingAdmissionPolicyBindingList(name, undefined, { urn });
            case "kubernetes:admissionregistration.k8s.io/v1beta1:ValidatingAdmissionPolicyBindingPatch":
                return new exports.ValidatingAdmissionPolicyBindingPatch(name, undefined, { urn });
            case "kubernetes:admissionregistration.k8s.io/v1beta1:ValidatingAdmissionPolicyList":
                return new exports.ValidatingAdmissionPolicyList(name, undefined, { urn });
            case "kubernetes:admissionregistration.k8s.io/v1beta1:ValidatingAdmissionPolicyPatch":
                return new exports.ValidatingAdmissionPolicyPatch(name, undefined, { urn });
            case "kubernetes:admissionregistration.k8s.io/v1beta1:ValidatingWebhookConfiguration":
                return new exports.ValidatingWebhookConfiguration(name, undefined, { urn });
            case "kubernetes:admissionregistration.k8s.io/v1beta1:ValidatingWebhookConfigurationList":
                return new exports.ValidatingWebhookConfigurationList(name, undefined, { urn });
            case "kubernetes:admissionregistration.k8s.io/v1beta1:ValidatingWebhookConfigurationPatch":
                return new exports.ValidatingWebhookConfigurationPatch(name, undefined, { urn });
            default:
                throw new Error(`unknown resource type ${type}`);
        }
    },
};
pulumi.runtime.registerResourceModule("kubernetes", "admissionregistration.k8s.io/v1beta1", _module);
