// @pulumi/kubernetes/admissionregistration/v1/validatingAdmissionPolicyBinding.js
"use strict";
// *** WARNING: this file was generated by pulumigen. ***
// *** Do not edit by hand unless you're certain you know what you are doing! ***
Object.defineProperty(exports, "__esModule", { value: true });
exports.ValidatingAdmissionPolicyBinding = void 0;
const pulumi = require("@pulumi/pulumi");
const utilities = require("../../utilities");
/**
 * ValidatingAdmissionPolicyBinding binds the ValidatingAdmissionPolicy with paramerized resources. ValidatingAdmissionPolicyBinding and parameter CRDs together define how cluster administrators configure policies for clusters.
 *
 * For a given admission request, each binding will cause its policy to be evaluated N times, where N is 1 for policies/bindings that don't use params, otherwise N is the number of parameters selected by the binding.
 *
 * The CEL expressions of a policy must have a computed CEL cost below the maximum CEL budget. Each evaluation of the policy is given an independent CEL cost budget. Adding/removing policies, bindings, or params can not affect whether a given (policy, binding, param) combination is within its own CEL budget.
 */
class ValidatingAdmissionPolicyBinding extends pulumi.CustomResource {
    /**
     * Get an existing ValidatingAdmissionPolicyBinding resource's state with the given name, ID, and optional extra
     * properties used to qualify the lookup.
     *
     * @param name The _unique_ name of the resulting resource.
     * @param id The _unique_ provider ID of the resource to lookup.
     * @param opts Optional settings to control the behavior of the CustomResource.
     */
    static get(name, id, opts) {
        return new ValidatingAdmissionPolicyBinding(name, undefined, Object.assign(Object.assign({}, opts), { id: id }));
    }
    /**
     * Returns true if the given object is an instance of ValidatingAdmissionPolicyBinding.  This is designed to work even
     * when multiple copies of the Pulumi SDK have been loaded into the same process.
     */
    static isInstance(obj) {
        if (obj === undefined || obj === null) {
            return false;
        }
        return obj['__pulumiType'] === ValidatingAdmissionPolicyBinding.__pulumiType;
    }
    /**
     * Create a ValidatingAdmissionPolicyBinding resource with the given unique name, arguments, and options.
     *
     * @param name The _unique_ name of the resource.
     * @param args The arguments to use to populate this resource's properties.
     * @param opts A bag of options that control this resource's behavior.
     */
    constructor(name, args, opts) {
        let resourceInputs = {};
        opts = opts || {};
        if (!opts.id) {
            resourceInputs["apiVersion"] = "admissionregistration.k8s.io/v1";
            resourceInputs["kind"] = "ValidatingAdmissionPolicyBinding";
            resourceInputs["metadata"] = args ? args.metadata : undefined;
            resourceInputs["spec"] = args ? args.spec : undefined;
        }
        else {
            resourceInputs["apiVersion"] = undefined /*out*/;
            resourceInputs["kind"] = undefined /*out*/;
            resourceInputs["metadata"] = undefined /*out*/;
            resourceInputs["spec"] = undefined /*out*/;
        }
        opts = pulumi.mergeOptions(utilities.resourceOptsDefaults(), opts);
        const aliasOpts = { aliases: [{ type: "kubernetes:admissionregistration.k8s.io/v1alpha1:ValidatingAdmissionPolicyBinding" }, { type: "kubernetes:admissionregistration.k8s.io/v1beta1:ValidatingAdmissionPolicyBinding" }] };
        opts = pulumi.mergeOptions(opts, aliasOpts);
        super(ValidatingAdmissionPolicyBinding.__pulumiType, name, resourceInputs, opts);
    }
}
exports.ValidatingAdmissionPolicyBinding = ValidatingAdmissionPolicyBinding;
/** @internal */
ValidatingAdmissionPolicyBinding.__pulumiType = 'kubernetes:admissionregistration.k8s.io/v1:ValidatingAdmissionPolicyBinding';
