// @pulumi/aws/workspacesweb/index.js
"use strict";
// *** WARNING: this file was generated by pulumi-language-nodejs. ***
// *** Do not edit by hand unless you're certain you know what you are doing! ***
Object.defineProperty(exports, "__esModule", { value: true });
exports.UserSettings = exports.UserAccessLoggingSettings = exports.NetworkSettings = exports.IpAccessSettings = exports.DataProtectionSettings = exports.BrowserSettings = void 0;
const pulumi = require("@pulumi/pulumi");
const utilities = require("../utilities");
exports.BrowserSettings = null;
utilities.lazyLoad(exports, ["BrowserSettings"], () => require("./browserSettings"));
exports.DataProtectionSettings = null;
utilities.lazyLoad(exports, ["DataProtectionSettings"], () => require("./dataProtectionSettings"));
exports.IpAccessSettings = null;
utilities.lazyLoad(exports, ["IpAccessSettings"], () => require("./ipAccessSettings"));
exports.NetworkSettings = null;
utilities.lazyLoad(exports, ["NetworkSettings"], () => require("./networkSettings"));
exports.UserAccessLoggingSettings = null;
utilities.lazyLoad(exports, ["UserAccessLoggingSettings"], () => require("./userAccessLoggingSettings"));
exports.UserSettings = null;
utilities.lazyLoad(exports, ["UserSettings"], () => require("./userSettings"));
const _module = {
    version: utilities.getVersion(),
    construct: (name, type, urn) => {
        switch (type) {
            case "aws:workspacesweb/browserSettings:BrowserSettings":
                return new exports.BrowserSettings(name, undefined, { urn });
            case "aws:workspacesweb/dataProtectionSettings:DataProtectionSettings":
                return new exports.DataProtectionSettings(name, undefined, { urn });
            case "aws:workspacesweb/ipAccessSettings:IpAccessSettings":
                return new exports.IpAccessSettings(name, undefined, { urn });
            case "aws:workspacesweb/networkSettings:NetworkSettings":
                return new exports.NetworkSettings(name, undefined, { urn });
            case "aws:workspacesweb/userAccessLoggingSettings:UserAccessLoggingSettings":
                return new exports.UserAccessLoggingSettings(name, undefined, { urn });
            case "aws:workspacesweb/userSettings:UserSettings":
                return new exports.UserSettings(name, undefined, { urn });
            default:
                throw new Error(`unknown resource type ${type}`);
        }
    },
};
pulumi.runtime.registerResourceModule("aws", "workspacesweb/browserSettings", _module);
pulumi.runtime.registerResourceModule("aws", "workspacesweb/dataProtectionSettings", _module);
pulumi.runtime.registerResourceModule("aws", "workspacesweb/ipAccessSettings", _module);
pulumi.runtime.registerResourceModule("aws", "workspacesweb/networkSettings", _module);
pulumi.runtime.registerResourceModule("aws", "workspacesweb/userAccessLoggingSettings", _module);
pulumi.runtime.registerResourceModule("aws", "workspacesweb/userSettings", _module);
