// @pulumi/kubernetes/resource/v1alpha2/resourceSlice.js
"use strict";
// *** WARNING: this file was generated by pulumigen. ***
// *** Do not edit by hand unless you're certain you know what you are doing! ***
Object.defineProperty(exports, "__esModule", { value: true });
exports.ResourceSlice = void 0;
const pulumi = require("@pulumi/pulumi");
const utilities = require("../../utilities");
/**
 * ResourceSlice provides information about available resources on individual nodes.
 */
class ResourceSlice extends pulumi.CustomResource {
    /**
     * Get an existing ResourceSlice resource's state with the given name, ID, and optional extra
     * properties used to qualify the lookup.
     *
     * @param name The _unique_ name of the resulting resource.
     * @param id The _unique_ provider ID of the resource to lookup.
     * @param opts Optional settings to control the behavior of the CustomResource.
     */
    static get(name, id, opts) {
        return new ResourceSlice(name, undefined, Object.assign(Object.assign({}, opts), { id: id }));
    }
    /**
     * Returns true if the given object is an instance of ResourceSlice.  This is designed to work even
     * when multiple copies of the Pulumi SDK have been loaded into the same process.
     */
    static isInstance(obj) {
        if (obj === undefined || obj === null) {
            return false;
        }
        return obj['__pulumiType'] === ResourceSlice.__pulumiType;
    }
    /**
     * Create a ResourceSlice resource with the given unique name, arguments, and options.
     *
     * @param name The _unique_ name of the resource.
     * @param args The arguments to use to populate this resource's properties.
     * @param opts A bag of options that control this resource's behavior.
     */
    constructor(name, args, opts) {
        let resourceInputs = {};
        opts = opts || {};
        if (!opts.id) {
            if ((!args || args.driverName === undefined) && !opts.urn) {
                throw new Error("Missing required property 'driverName'");
            }
            resourceInputs["apiVersion"] = "resource.k8s.io/v1alpha2";
            resourceInputs["driverName"] = args ? args.driverName : undefined;
            resourceInputs["kind"] = "ResourceSlice";
            resourceInputs["metadata"] = args ? args.metadata : undefined;
            resourceInputs["namedResources"] = args ? args.namedResources : undefined;
            resourceInputs["nodeName"] = args ? args.nodeName : undefined;
        }
        else {
            resourceInputs["apiVersion"] = undefined /*out*/;
            resourceInputs["driverName"] = undefined /*out*/;
            resourceInputs["kind"] = undefined /*out*/;
            resourceInputs["metadata"] = undefined /*out*/;
            resourceInputs["namedResources"] = undefined /*out*/;
            resourceInputs["nodeName"] = undefined /*out*/;
        }
        opts = pulumi.mergeOptions(utilities.resourceOptsDefaults(), opts);
        const aliasOpts = { aliases: [{ type: "kubernetes:resource.k8s.io/v1alpha3:ResourceSlice" }, { type: "kubernetes:resource.k8s.io/v1beta1:ResourceSlice" }, { type: "kubernetes:resource.k8s.io/v1beta2:ResourceSlice" }] };
        opts = pulumi.mergeOptions(opts, aliasOpts);
        super(ResourceSlice.__pulumiType, name, resourceInputs, opts);
    }
}
exports.ResourceSlice = ResourceSlice;
/** @internal */
ResourceSlice.__pulumiType = 'kubernetes:resource.k8s.io/v1alpha2:ResourceSlice';
