// @pulumi/kubernetes/resource/v1alpha2/resourceClassParameters.js
"use strict";
// *** WARNING: this file was generated by pulumigen. ***
// *** Do not edit by hand unless you're certain you know what you are doing! ***
Object.defineProperty(exports, "__esModule", { value: true });
exports.ResourceClassParameters = void 0;
const pulumi = require("@pulumi/pulumi");
const utilities = require("../../utilities");
/**
 * ResourceClassParameters defines resource requests for a ResourceClass in an in-tree format understood by Kubernetes.
 */
class ResourceClassParameters extends pulumi.CustomResource {
    /**
     * Get an existing ResourceClassParameters resource's state with the given name, ID, and optional extra
     * properties used to qualify the lookup.
     *
     * @param name The _unique_ name of the resulting resource.
     * @param id The _unique_ provider ID of the resource to lookup.
     * @param opts Optional settings to control the behavior of the CustomResource.
     */
    static get(name, id, opts) {
        return new ResourceClassParameters(name, undefined, Object.assign(Object.assign({}, opts), { id: id }));
    }
    /**
     * Returns true if the given object is an instance of ResourceClassParameters.  This is designed to work even
     * when multiple copies of the Pulumi SDK have been loaded into the same process.
     */
    static isInstance(obj) {
        if (obj === undefined || obj === null) {
            return false;
        }
        return obj['__pulumiType'] === ResourceClassParameters.__pulumiType;
    }
    /**
     * Create a ResourceClassParameters resource with the given unique name, arguments, and options.
     *
     * @param name The _unique_ name of the resource.
     * @param args The arguments to use to populate this resource's properties.
     * @param opts A bag of options that control this resource's behavior.
     */
    constructor(name, args, opts) {
        let resourceInputs = {};
        opts = opts || {};
        if (!opts.id) {
            resourceInputs["apiVersion"] = "resource.k8s.io/v1alpha2";
            resourceInputs["filters"] = args ? args.filters : undefined;
            resourceInputs["generatedFrom"] = args ? args.generatedFrom : undefined;
            resourceInputs["kind"] = "ResourceClassParameters";
            resourceInputs["metadata"] = args ? args.metadata : undefined;
            resourceInputs["vendorParameters"] = args ? args.vendorParameters : undefined;
        }
        else {
            resourceInputs["apiVersion"] = undefined /*out*/;
            resourceInputs["filters"] = undefined /*out*/;
            resourceInputs["generatedFrom"] = undefined /*out*/;
            resourceInputs["kind"] = undefined /*out*/;
            resourceInputs["metadata"] = undefined /*out*/;
            resourceInputs["vendorParameters"] = undefined /*out*/;
        }
        opts = pulumi.mergeOptions(utilities.resourceOptsDefaults(), opts);
        super(ResourceClassParameters.__pulumiType, name, resourceInputs, opts);
    }
}
exports.ResourceClassParameters = ResourceClassParameters;
/** @internal */
ResourceClassParameters.__pulumiType = 'kubernetes:resource.k8s.io/v1alpha2:ResourceClassParameters';
