// @pulumi/aws/vpc/index.js
"use strict";
// *** WARNING: this file was generated by pulumi-language-nodejs. ***
// *** Do not edit by hand unless you're certain you know what you are doing! ***
Object.defineProperty(exports, "__esModule", { value: true });
exports.SecurityGroupVpcAssociation = exports.SecurityGroupIngressRule = exports.SecurityGroupEgressRule = exports.RouteServerVpcAssociation = exports.RouteServerPropagation = exports.RouteServerPeer = exports.RouteServerEndpoint = exports.RouteServer = exports.getVpcEndpointAssociationsOutput = exports.getVpcEndpointAssociations = exports.getSecurityGroupRulesOutput = exports.getSecurityGroupRules = exports.getSecurityGroupRuleOutput = exports.getSecurityGroupRule = exports.EndpointServicePrivateDnsVerification = exports.EndpointPrivateDns = void 0;
const pulumi = require("@pulumi/pulumi");
const utilities = require("../utilities");
exports.EndpointPrivateDns = null;
utilities.lazyLoad(exports, ["EndpointPrivateDns"], () => require("./endpointPrivateDns"));
exports.EndpointServicePrivateDnsVerification = null;
utilities.lazyLoad(exports, ["EndpointServicePrivateDnsVerification"], () => require("./endpointServicePrivateDnsVerification"));
exports.getSecurityGroupRule = null;
exports.getSecurityGroupRuleOutput = null;
utilities.lazyLoad(exports, ["getSecurityGroupRule", "getSecurityGroupRuleOutput"], () => require("./getSecurityGroupRule"));
exports.getSecurityGroupRules = null;
exports.getSecurityGroupRulesOutput = null;
utilities.lazyLoad(exports, ["getSecurityGroupRules", "getSecurityGroupRulesOutput"], () => require("./getSecurityGroupRules"));
exports.getVpcEndpointAssociations = null;
exports.getVpcEndpointAssociationsOutput = null;
utilities.lazyLoad(exports, ["getVpcEndpointAssociations", "getVpcEndpointAssociationsOutput"], () => require("./getVpcEndpointAssociations"));
exports.RouteServer = null;
utilities.lazyLoad(exports, ["RouteServer"], () => require("./routeServer"));
exports.RouteServerEndpoint = null;
utilities.lazyLoad(exports, ["RouteServerEndpoint"], () => require("./routeServerEndpoint"));
exports.RouteServerPeer = null;
utilities.lazyLoad(exports, ["RouteServerPeer"], () => require("./routeServerPeer"));
exports.RouteServerPropagation = null;
utilities.lazyLoad(exports, ["RouteServerPropagation"], () => require("./routeServerPropagation"));
exports.RouteServerVpcAssociation = null;
utilities.lazyLoad(exports, ["RouteServerVpcAssociation"], () => require("./routeServerVpcAssociation"));
exports.SecurityGroupEgressRule = null;
utilities.lazyLoad(exports, ["SecurityGroupEgressRule"], () => require("./securityGroupEgressRule"));
exports.SecurityGroupIngressRule = null;
utilities.lazyLoad(exports, ["SecurityGroupIngressRule"], () => require("./securityGroupIngressRule"));
exports.SecurityGroupVpcAssociation = null;
utilities.lazyLoad(exports, ["SecurityGroupVpcAssociation"], () => require("./securityGroupVpcAssociation"));
const _module = {
    version: utilities.getVersion(),
    construct: (name, type, urn) => {
        switch (type) {
            case "aws:vpc/endpointPrivateDns:EndpointPrivateDns":
                return new exports.EndpointPrivateDns(name, undefined, { urn });
            case "aws:vpc/endpointServicePrivateDnsVerification:EndpointServicePrivateDnsVerification":
                return new exports.EndpointServicePrivateDnsVerification(name, undefined, { urn });
            case "aws:vpc/routeServer:RouteServer":
                return new exports.RouteServer(name, undefined, { urn });
            case "aws:vpc/routeServerEndpoint:RouteServerEndpoint":
                return new exports.RouteServerEndpoint(name, undefined, { urn });
            case "aws:vpc/routeServerPeer:RouteServerPeer":
                return new exports.RouteServerPeer(name, undefined, { urn });
            case "aws:vpc/routeServerPropagation:RouteServerPropagation":
                return new exports.RouteServerPropagation(name, undefined, { urn });
            case "aws:vpc/routeServerVpcAssociation:RouteServerVpcAssociation":
                return new exports.RouteServerVpcAssociation(name, undefined, { urn });
            case "aws:vpc/securityGroupEgressRule:SecurityGroupEgressRule":
                return new exports.SecurityGroupEgressRule(name, undefined, { urn });
            case "aws:vpc/securityGroupIngressRule:SecurityGroupIngressRule":
                return new exports.SecurityGroupIngressRule(name, undefined, { urn });
            case "aws:vpc/securityGroupVpcAssociation:SecurityGroupVpcAssociation":
                return new exports.SecurityGroupVpcAssociation(name, undefined, { urn });
            default:
                throw new Error(`unknown resource type ${type}`);
        }
    },
};
pulumi.runtime.registerResourceModule("aws", "vpc/endpointPrivateDns", _module);
pulumi.runtime.registerResourceModule("aws", "vpc/endpointServicePrivateDnsVerification", _module);
pulumi.runtime.registerResourceModule("aws", "vpc/routeServer", _module);
pulumi.runtime.registerResourceModule("aws", "vpc/routeServerEndpoint", _module);
pulumi.runtime.registerResourceModule("aws", "vpc/routeServerPeer", _module);
pulumi.runtime.registerResourceModule("aws", "vpc/routeServerPropagation", _module);
pulumi.runtime.registerResourceModule("aws", "vpc/routeServerVpcAssociation", _module);
pulumi.runtime.registerResourceModule("aws", "vpc/securityGroupEgressRule", _module);
pulumi.runtime.registerResourceModule("aws", "vpc/securityGroupIngressRule", _module);
pulumi.runtime.registerResourceModule("aws", "vpc/securityGroupVpcAssociation", _module);
