// @pulumi/kubernetes/batch/v1/index.js
"use strict";
// *** WARNING: this file was generated by pulumigen. ***
// *** Do not edit by hand unless you're certain you know what you are doing! ***
Object.defineProperty(exports, "__esModule", { value: true });
exports.JobPatch = exports.JobList = exports.Job = exports.CronJobPatch = exports.CronJobList = exports.CronJob = void 0;
const pulumi = require("@pulumi/pulumi");
const utilities = require("../../utilities");
exports.CronJob = null;
utilities.lazyLoad(exports, ["CronJob"], () => require("./cronJob"));
exports.CronJobList = null;
utilities.lazyLoad(exports, ["CronJobList"], () => require("./cronJobList"));
exports.CronJobPatch = null;
utilities.lazyLoad(exports, ["CronJobPatch"], () => require("./cronJobPatch"));
exports.Job = null;
utilities.lazyLoad(exports, ["Job"], () => require("./job"));
exports.JobList = null;
utilities.lazyLoad(exports, ["JobList"], () => require("./jobList"));
exports.JobPatch = null;
utilities.lazyLoad(exports, ["JobPatch"], () => require("./jobPatch"));
const _module = {
    version: utilities.getVersion(),
    construct: (name, type, urn) => {
        switch (type) {
            case "kubernetes:batch/v1:CronJob":
                return new exports.CronJob(name, undefined, { urn });
            case "kubernetes:batch/v1:CronJobList":
                return new exports.CronJobList(name, undefined, { urn });
            case "kubernetes:batch/v1:CronJobPatch":
                return new exports.CronJobPatch(name, undefined, { urn });
            case "kubernetes:batch/v1:Job":
                return new exports.Job(name, undefined, { urn });
            case "kubernetes:batch/v1:JobList":
                return new exports.JobList(name, undefined, { urn });
            case "kubernetes:batch/v1:JobPatch":
                return new exports.JobPatch(name, undefined, { urn });
            default:
                throw new Error(`unknown resource type ${type}`);
        }
    },
};
pulumi.runtime.registerResourceModule("kubernetes", "batch/v1", _module);
