// @pulumi/eks/cni.js
"use strict";
// Copyright 2016-2019, Pulumi Corporation.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.
Object.defineProperty(exports, "__esModule", { value: true });
exports.VpcCni = void 0;
const pulumi = require("@pulumi/pulumi");
const utilities = require("./utilities");
// Dynamic providers don't currently work well with multi-language components [1]. To workaround this, the
// dynamic providers in this library have been ported to be custom resources implemented by the new provider
// plugin. An alias is used and care has been taken in the implementation inside the provider to avoid any
// diffs for any existing stacks that were created using the old dynamic provider.
// [1] https://github.com/pulumi/pulumi/issues/5455
/**
 * VpcCni manages the configuration of the Amazon VPC CNI plugin for Kubernetes by applying its YAML chart. Once Pulumi is
 * able to programatically manage existing infrastructure, we can replace this with a real k8s resource.
 */
class VpcCni extends pulumi.CustomResource {
    constructor(name, kubeconfig, args, opts) {
        const defaultOpts = {
            version: utilities.getVersion(),
            // This was previously implemented as a dynamic provider, so alias the old type.
            aliases: [{ type: "pulumi-nodejs:dynamic:Resource" }],
        };
        opts = pulumi.mergeOptions(opts, defaultOpts);
        super("eks:index:VpcCni", name, {
            kubeconfig: kubeconfig,
            nodePortSupport: args === null || args === void 0 ? void 0 : args.nodePortSupport,
            customNetworkConfig: args === null || args === void 0 ? void 0 : args.customNetworkConfig,
            externalSnat: args === null || args === void 0 ? void 0 : args.externalSnat,
            warmEniTarget: args === null || args === void 0 ? void 0 : args.warmEniTarget,
            warmIpTarget: args === null || args === void 0 ? void 0 : args.warmIpTarget,
            enablePrefixDelegation: args === null || args === void 0 ? void 0 : args.enablePrefixDelegation,
            enableIpv6: args === null || args === void 0 ? void 0 : args.enableIpv6,
            warmPrefixTarget: args === null || args === void 0 ? void 0 : args.warmPrefixTarget,
            logLevel: args === null || args === void 0 ? void 0 : args.logLevel,
            logFile: args === null || args === void 0 ? void 0 : args.logFile,
            image: args === null || args === void 0 ? void 0 : args.image,
            nodeAgentImage: args === null || args === void 0 ? void 0 : args.nodeAgentImage,
            initImage: args === null || args === void 0 ? void 0 : args.initImage,
            eniConfigLabelDef: args === null || args === void 0 ? void 0 : args.eniConfigLabelDef,
            pluginLogLevel: args === null || args === void 0 ? void 0 : args.pluginLogLevel,
            pluginLogFile: args === null || args === void 0 ? void 0 : args.pluginLogFile,
            enablePodEni: args === null || args === void 0 ? void 0 : args.enablePodEni,
            disableTcpEarlyDemux: args === null || args === void 0 ? void 0 : args.disableTcpEarlyDemux,
            cniConfigureRpfilter: args === null || args === void 0 ? void 0 : args.cniConfigureRpfilter,
            cniCustomNetworkCfg: args === null || args === void 0 ? void 0 : args.cniCustomNetworkCfg,
            cniExternalSnat: args === null || args === void 0 ? void 0 : args.cniExternalSnat,
            securityContextPrivileged: args === null || args === void 0 ? void 0 : args.securityContextPrivileged,
        }, opts);
    }
}
exports.VpcCni = VpcCni;
