// @pulumi/aws/workspaces/webDataProtectionSettings.js
"use strict";
// *** WARNING: this file was generated by pulumi-language-nodejs. ***
// *** Do not edit by hand unless you're certain you know what you are doing! ***
Object.defineProperty(exports, "__esModule", { value: true });
exports.WebDataProtectionSettings = void 0;
const pulumi = require("@pulumi/pulumi");
const utilities = require("../utilities");
/**
 * Resource for managing an AWS WorkSpaces Web Data Protection Settings resource. Once associated with a web portal, data protection settings control how sensitive information is redacted in streaming sessions.
 *
 * ## Example Usage
 *
 * ### Basic Usage
 *
 * ```typescript
 * import * as pulumi from "@pulumi/pulumi";
 * import * as aws from "@pulumi/aws";
 *
 * const example = new aws.workspacesweb.DataProtectionSettings("example", {displayName: "example"});
 * ```
 *
 * ### With Inline Redaction Configuration
 *
 * ```typescript
 * import * as pulumi from "@pulumi/pulumi";
 * import * as aws from "@pulumi/aws";
 *
 * const example = new aws.workspacesweb.DataProtectionSettings("example", {
 *     displayName: "example",
 *     description: "Example data protection settings",
 *     inlineRedactionConfiguration: {
 *         globalConfidenceLevel: 2,
 *         globalEnforcedUrls: ["https://example.com"],
 *         inlineRedactionPatterns: [{
 *             builtInPatternId: "ssn",
 *             confidenceLevel: 3,
 *             redactionPlaceHolders: [{
 *                 redactionPlaceHolderType: "CustomText",
 *                 redactionPlaceHolderText: "REDACTED",
 *             }],
 *         }],
 *     },
 * });
 * ```
 *
 * ### Complete Example
 *
 * ```typescript
 * import * as pulumi from "@pulumi/pulumi";
 * import * as aws from "@pulumi/aws";
 *
 * const example = new aws.kms.Key("example", {
 *     description: "KMS key for WorkSpaces Web Data Protection Settings",
 *     deletionWindowInDays: 7,
 * });
 * const exampleDataProtectionSettings = new aws.workspacesweb.DataProtectionSettings("example", {
 *     displayName: "example-complete",
 *     description: "Complete example data protection settings",
 *     customerManagedKey: example.arn,
 *     additionalEncryptionContext: {
 *         Environment: "Production",
 *     },
 *     inlineRedactionConfiguration: {
 *         globalConfidenceLevel: 2,
 *         globalEnforcedUrls: [
 *             "https://example.com",
 *             "https://test.example.com",
 *         ],
 *         globalExemptUrls: ["https://exempt.example.com"],
 *         inlineRedactionPatterns: [
 *             {
 *                 builtInPatternId: "ssn",
 *                 confidenceLevel: 3,
 *                 enforcedUrls: ["https://pattern1.example.com"],
 *                 exemptUrls: ["https://exempt-pattern1.example.com"],
 *                 redactionPlaceHolders: [{
 *                     redactionPlaceHolderType: "CustomText",
 *                     redactionPlaceHolderText: "REDACTED-SSN",
 *                 }],
 *             },
 *             {
 *                 customPattern: {
 *                     patternName: "CustomPattern",
 *                     patternRegex: "/\\d{3}-\\d{2}-\\d{4}/g",
 *                     keywordRegex: "/SSN|Social Security/gi",
 *                     patternDescription: "Custom SSN pattern",
 *                 },
 *                 redactionPlaceHolders: [{
 *                     redactionPlaceHolderType: "CustomText",
 *                     redactionPlaceHolderText: "REDACTED-CUSTOM",
 *                 }],
 *             },
 *         ],
 *     },
 *     tags: {
 *         Name: "example-data-protection-settings",
 *     },
 * });
 * ```
 *
 * ## Import
 *
 * Using `pulumi import`, import WorkSpaces Web Data Protection Settings using the `data_protection_settings_arn`. For example:
 *
 * ```sh
 * $ pulumi import aws:workspaces/webDataProtectionSettings:WebDataProtectionSettings example arn:aws:workspaces-web:us-west-2:123456789012:dataprotectionsettings/abcdef12345
 * ```
 *
 * @deprecated aws.workspaces/webdataprotectionsettings.WebDataProtectionSettings has been deprecated in favor of aws.workspacesweb/dataprotectionsettings.DataProtectionSettings
 */
class WebDataProtectionSettings extends pulumi.CustomResource {
    /**
     * Get an existing WebDataProtectionSettings resource's state with the given name, ID, and optional extra
     * properties used to qualify the lookup.
     *
     * @param name The _unique_ name of the resulting resource.
     * @param id The _unique_ provider ID of the resource to lookup.
     * @param state Any extra arguments used during the lookup.
     * @param opts Optional settings to control the behavior of the CustomResource.
     */
    static get(name, id, state, opts) {
        pulumi.log.warn("WebDataProtectionSettings is deprecated: aws.workspaces/webdataprotectionsettings.WebDataProtectionSettings has been deprecated in favor of aws.workspacesweb/dataprotectionsettings.DataProtectionSettings");
        return new WebDataProtectionSettings(name, state, Object.assign(Object.assign({}, opts), { id: id }));
    }
    /**
     * Returns true if the given object is an instance of WebDataProtectionSettings.  This is designed to work even
     * when multiple copies of the Pulumi SDK have been loaded into the same process.
     */
    static isInstance(obj) {
        if (obj === undefined || obj === null) {
            return false;
        }
        return obj['__pulumiType'] === WebDataProtectionSettings.__pulumiType;
    }
    /** @deprecated aws.workspaces/webdataprotectionsettings.WebDataProtectionSettings has been deprecated in favor of aws.workspacesweb/dataprotectionsettings.DataProtectionSettings */
    constructor(name, argsOrState, opts) {
        pulumi.log.warn("WebDataProtectionSettings is deprecated: aws.workspaces/webdataprotectionsettings.WebDataProtectionSettings has been deprecated in favor of aws.workspacesweb/dataprotectionsettings.DataProtectionSettings");
        let resourceInputs = {};
        opts = opts || {};
        if (opts.id) {
            const state = argsOrState;
            resourceInputs["additionalEncryptionContext"] = state ? state.additionalEncryptionContext : undefined;
            resourceInputs["associatedPortalArns"] = state ? state.associatedPortalArns : undefined;
            resourceInputs["customerManagedKey"] = state ? state.customerManagedKey : undefined;
            resourceInputs["dataProtectionSettingsArn"] = state ? state.dataProtectionSettingsArn : undefined;
            resourceInputs["description"] = state ? state.description : undefined;
            resourceInputs["displayName"] = state ? state.displayName : undefined;
            resourceInputs["inlineRedactionConfiguration"] = state ? state.inlineRedactionConfiguration : undefined;
            resourceInputs["tags"] = state ? state.tags : undefined;
            resourceInputs["tagsAll"] = state ? state.tagsAll : undefined;
        }
        else {
            const args = argsOrState;
            if ((!args || args.displayName === undefined) && !opts.urn) {
                throw new Error("Missing required property 'displayName'");
            }
            resourceInputs["additionalEncryptionContext"] = args ? args.additionalEncryptionContext : undefined;
            resourceInputs["customerManagedKey"] = args ? args.customerManagedKey : undefined;
            resourceInputs["description"] = args ? args.description : undefined;
            resourceInputs["displayName"] = args ? args.displayName : undefined;
            resourceInputs["inlineRedactionConfiguration"] = args ? args.inlineRedactionConfiguration : undefined;
            resourceInputs["tags"] = args ? args.tags : undefined;
            resourceInputs["associatedPortalArns"] = undefined /*out*/;
            resourceInputs["dataProtectionSettingsArn"] = undefined /*out*/;
            resourceInputs["tagsAll"] = undefined /*out*/;
        }
        opts = pulumi.mergeOptions(utilities.resourceOptsDefaults(), opts);
        super(WebDataProtectionSettings.__pulumiType, name, resourceInputs, opts);
    }
}
exports.WebDataProtectionSettings = WebDataProtectionSettings;
/** @internal */
WebDataProtectionSettings.__pulumiType = 'aws:workspaces/webDataProtectionSettings:WebDataProtectionSettings';
